Userland/Libraries/LibGfx/ImageFormats/JPEG2000Loader.cpp

```diff
@@ -2162,78 +2163,93 @@ static ErrorOr<void> convert_to_bitmap(JPEG2000LoadingContext& context)
-        if (context.channel_definition_box->channels.size() != context.siz.components.size())
-            return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Channel definition box channel count doesn't match component count");
-
-        Vector<bool, 4> channel_used;
-        channel_used.resize(context.channel_definition_box->channels.size());
-
-        for (auto channel : context.channel_definition_box->channels) {
-            if (channel.channel_index >= context.siz.components.size())
-                return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Channel definition box channel index out of range");
-            if (channel_used[channel.channel_index])
-                return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Channel definition box channel index used multiple times");
-            channel_used[channel.channel_index] = true;
-
-            if (channel.channel_type != ISOBMFF::JPEG2000ChannelDefinitionBox::Channel::Type::Color
-                && channel.channel_type != ISOBMFF::JPEG2000ChannelDefinitionBox::Channel::Type::Opacity) {
-                return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Only color and opacity channels supported yet");
+    if (context.palette_box.has_value() && context.options.palette_handling == JPEG2000DecoderOptions::PaletteHandling::PaletteIndicesAsGrayscale) {
+        for (auto& tile : context.tiles) {
+            if (tile.components.size() != 1)
+                return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Palette indices as grayscale require single component");
+            for (auto sample : tile.components[0].samples) {
+                // The JPEG2000 spec allows palette indices up to 1023, but the PDF spec says that JPEG2000 images
+                // embedded in PDFs must have indices that fit in a one byte.
+                if (sample < 0 || sample >= 256)
+                    return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Palette indices out of range");
             }
-            if (channel.channel_type == ISOBMFF::JPEG2000ChannelDefinitionBox::Channel::Type::Color) {
-                if (channel.channel_index + 1 != channel.channel_association)
-                    return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Only unshuffled color channel indices supported yet");
-            } else {
-                VERIFY(channel.channel_type == ISOBMFF::JPEG2000ChannelDefinitionBox::Channel::Type::Opacity);
-                if (channel.channel_index != context.siz.components.size() - 1)
-                    return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Only opacity channel as last channel supported yet");
-                if (channel.channel_association != 0)
-                    return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Only full opacity channel supported yet");
-                has_alpha = true;
+        }
+    } else {
```

Comment:
The diff doesn't make this very visible, but everything in the `else` is everything that was on the LHS, just indented one level.